bin/cron/applab_datasets/spotify

```diff
@@ -4,100 +4,132 @@ abort 'Script already running' unless only_one_running?(__FILE__)
-TOP_200_WORLD_URL = 'https://spotifycharts.com/regional/global/daily/latest/download'
-VIRAL_50_US_URL = 'https://spotifycharts.com/viral/us/daily/latest/download'
-VIRAL_50_WORLD_URL = 'https://spotifycharts.com/viral/global/daily/latest/download'
 PROFANITY_LOCALE = 'en-US'
 
-def get_top200_data(url)
-  records = {}
-  columns = ['id', 'Track Name', 'Artist', 'Position', 'Streams', 'URL']
-  response = Net::HTTP.get_response(URI(url))
-  return unless response.code == '200'
-  id = 1
-  response.body.force_encoding('utf-8')
-  csv = CSV.new(response.body)
-  lines = csv.read
-  lines.shift # first line just contains a note
-  lines.shift # second line is column names
-  regex = generate_profanity_regex(lines)
-  lines.each do |line|
-    next if regex&.match?(get_string_to_check(line)) # skip record if it contains profanity
+CHART_TYPE = {
+  TOP: 'TOP',
+  VIRAL: 'VIRAL',
+}
 
-    record = OpenStruct.new
-    record.id = id
-    record.Position = line[0].to_i
-    record['Track Name'] = line[1]
-    record.Artist = line[2]
-    record.Streams = line[3].to_i
-    record.URL = line[4]
+def get_token
+  client_id = CDO.spotify_api_client_id
+  client_secret = CDO.spotify_api_client_secret
 
-    records[id] = record.to_h.to_json
-    id += 1
+  begin
+    url = "https://accounts.spotify.com/api/token"
+    headers = {
+      'authorization' => 'Basic ' + Base64.strict_encode64(client_id + ':' + client_secret),
+      'content-type' => 'application/x-www-form-urlencoded',
+    }
+    form = {
+      'grant_type' => 'client_credentials'
+    }
+    auth_response = JSON.parse(RestClient.post(url, form, headers))
+    $token = auth_response['access_token']
+  rescue RestClient::Exception => e
+    Honeybadger.notify(
+      e,
+      error_message: "Failed to retrieve OAuth token from Spotify for use in App Lab Datasets.",
+      context: {
+        tenant_id: client_id,
+        url: url
+      }
+    )
   end
-  return records, columns
 end
 
-def get_viral50_data(url)
-  records = {}
+def get_playlist_data(chart_type, playlist_id)
+  # Filters for the query: a comma-separated list of the fields to return
+  field_query = 'fields=tracks.items(track(name,href,artists,popularity))'
+
+  begin
+    url = "#{BASE_URL}/playlists/#{playlist_id}?#{field_query}"
+    response = RestClient.get(url, {authorization: "Bearer #{$token}"})
+  rescue RestClient::Exception => e
+    Honeybadger.notify(
+      e,
+      error_message: "Failed to retrieve playlist data from Spotify for use in App Lab Datasets.",
+      context: {
+        playlist_id: playlist_id,
+      }
+    )
+    return
+  end
+
+  items = JSON.parse(response)['tracks']['items']
   columns = ['id', 'Track Name', 'Artist', 'Position', 'URL']
-  response = Net::HTTP.get_response(URI(url))
-  return unless response.code == '200'
+  if chart_type === CHART_TYPE[:TOP]
+    columns.insert(4, 'Popularity')   # Only include Popularity for Top Charts
+  end
```

Comment:
It's been a while since I've worked with ruby so definitely open to any ideas / suggestions of a better way to handle conditionally adding in this column!